Error requiring the IEditorTracker

Error requiring the IEditorTracker

## Description

```
node_modules/@jupyterlab/fileeditor/lib/toc/factory.d.ts:24:5 - error TS2416: Property 'createNew' in type 'EditorTableOfContentsFactory' is not assignable to the same property in base type 'TableOfContentsFactory<IDocumentWidget<FileEditor, IModel>>'.
  Type '(widget: IDocumentWidget<FileEditor, IModel>, configuration?: IConfig | undefined) => TableOfContentsModel<IEditorHeading, IDocumentWidget<...>>' is not assignable to type '(widget: IDocumentWidget<FileEditor, IModel>, configuration?: IConfig | undefined) => TableOfContentsModel<IHeading, IDocumentWidget<FileEditor, IModel>>'.
    Call signature return types 'TableOfContentsModel<IEditorHeading, IDocumentWidget<FileEditor, IModel>>' and 'TableOfContentsModel<IHeading, IDocumentWidget<FileEditor, IModel>>' are incompatible.
      The types of 'activeHeadingChanged' are incompatible between these types.
        Type 'ISignal<IModel<IEditorHeading>, IEditorHeading | null>' is not assignable to type 'ISignal<IModel<IHeading>, IHeading | null>'.
          Type 'IModel<IEditorHeading>' is not assignable to type 'IModel<IHeading>'.
            Types of property 'toggleCollapse' are incompatible.
              Type '(options: { heading?: IEditorHeading | undefined; collapsed?: boolean | undefined; }) => void' is not assignable to type '(options: { heading?: IHeading | undefined; collapsed?: boolean | undefined; }) => void'.
                Types of parameters 'options' and 'options' are incompatible.
                  Type '{ heading?: IHeading | undefined; collapsed?: boolean | undefined; }' is not assignable to type '{ heading?: IEditorHeading | undefined; collapsed?: boolean | undefined; }'.
                    Types of property 'heading' are incompatible.
                      Type 'IHeading | undefined' is not assignable to type 'IEditorHeading | undefined'.
                        Type 'IHeading' is not assignable to type 'IEditorHeading'.

24     createNew(widget: IDocumentWidget<FileEditor, DocumentRegistry.IModel>, configuration?: TableOfContents.IConfig): TableOfContentsModel<IEditorHeading, IDocumentWidget<FileEditor, DocumentRegistry.IModel>>;
```
I think the error comes from these lines:

- [ISignal\<TableOfContents.IModel\<H>, H | null>](https://github.com/jupyterlab/jupyterlab/blob/acb87d51ed0c489fb5a8f6781059024cf62f255b/packages/toc/src/model.ts#L54)
- [ISignal\<TableOfContents.IModel\<H>, H | null>](https://github.com/jupyterlab/jupyterlab/blob/acb87d51ed0c489fb5a8f6781059024cf62f255b/packages/toc/src/model.ts#L61)
- [ISignal\<TableOfContents.IModel\<H>, void>](https://github.com/jupyterlab/jupyterlab/blob/acb87d51ed0c489fb5a8f6781059024cf62f255b/packages/toc/src/model.ts#L93)

The public API uses `TableOfContents.IModel<H>` while each property uses `TableOfContentsModel<H, T>`. See:

- https://github.com/jupyterlab/jupyterlab/blob/acb87d51ed0c489fb5a8f6781059024cf62f255b/packages/toc/src/model.ts#L248
- https://github.com/jupyterlab/jupyterlab/blob/acb87d51ed0c489fb5a8f6781059024cf62f255b/packages/toc/src/model.ts#L245
- https://github.com/jupyterlab/jupyterlab/blob/acb87d51ed0c489fb5a8f6781059024cf62f255b/packages/toc/src/model.ts#LL244C41-L244C67
- https://github.com/jupyterlab/jupyterlab/blob/acb87d51ed0c489fb5a8f6781059024cf62f255b/packages/toc/src/model.ts#L29-L36

## Reproduce
Create a new extension for lab 4 from the [cookiecutter](https://github.com/jupyterlab/extension-cookiecutter-ts/tree/4.0) and use the following sniped to create a plugin that requires the IEditorTracker.
```
import { IEditorTracker } from '@jupyterlab/fileeditor';
const plugin: JupyterFrontEndPlugin<void> = {
  id: 'test_ext:plugin',
  description: 'x',
  autoStart: true,
  requires: [IEditorTracker],
  activate: (app: JupyterFrontEnd, tracker: IEditorTracker) => {
    console.log('JupyterLab extension test_ext is activated!');

    tracker.currentChanged.connect(() => console.debug("Doc changed"));
  }
};
```

## Expected behavior


## Context

JupyterLab version: 4.0.0rc0


cc @fcollonval

--- a/packages/fileeditor/src/toc/factory.ts
+++ b/packages/fileeditor/src/toc/factory.ts
@@ -23,7 +23,8 @@ export interface IEditorHeading extends TableOfContents.IHeading {
  * Base table of contents model factory for file editor
  */
 export abstract class EditorTableOfContentsFactory extends TableOfContentsFactory<
-  IDocumentWidget<FileEditor>
+  IDocumentWidget<FileEditor>,
+  IEditorHeading
 > {
   /**
    * Create a new table of contents model for the widget
@@ -39,13 +40,7 @@ export abstract class EditorTableOfContentsFactory extends TableOfContentsFactor
     IEditorHeading,
     IDocumentWidget<FileEditor, DocumentRegistry.IModel>
   > {
-    const model = super.createNew(
-      widget,
-      configuration
-    ) as TableOfContentsModel<
-      IEditorHeading,
-      IDocumentWidget<FileEditor, DocumentRegistry.IModel>
-    >;
+    const model = super.createNew(widget, configuration);
 
     const onActiveHeadingChanged = (
       model: TableOfContentsModel<

--- a/packages/toc/src/factory.ts
+++ b/packages/toc/src/factory.ts
@@ -18,8 +18,10 @@ const RENDER_TIMEOUT = 1000;
 /**
  * Abstract table of contents model factory for IDocumentWidget.
  */
-export abstract class TableOfContentsFactory<W extends IDocumentWidget>
-  implements TableOfContents.IFactory<W>
+export abstract class TableOfContentsFactory<
+  W extends IDocumentWidget,
+  H extends TableOfContents.IHeading = TableOfContents.IHeading
+> implements TableOfContents.IFactory<W, H>
 {
   /**
    * Constructor
@@ -52,7 +54,7 @@ export abstract class TableOfContentsFactory<W extends IDocumentWidget>
   createNew(
     widget: W,
     configuration?: TableOfContents.IConfig
-  ): TableOfContentsModel<TableOfContents.IHeading, W> {
+  ): TableOfContentsModel<H, W> {
     const model = this._createNew(widget, configuration);
 
     const context = widget.context;
@@ -102,5 +104,5 @@ export abstract class TableOfContentsFactory<W extends IDocumentWidget>
   protected abstract _createNew(
     widget: W,
     configuration?: TableOfContents.IConfig
-  ): TableOfContentsModel<TableOfContents.IHeading, W>;
+  ): TableOfContentsModel<H, W>;
 }

--- a/packages/toc/src/tokens.ts
+++ b/packages/toc/src/tokens.ts
@@ -73,7 +73,10 @@ export namespace TableOfContents {
   /**
    * Table of content model factory interface
    */
-  export interface IFactory<W extends Widget = Widget> {
+  export interface IFactory<
+    W extends Widget = Widget,
+    H extends IHeading = IHeading
+  > {
     /**
      * Whether the factory can handle the widget or not.
      *
@@ -92,7 +95,7 @@ export namespace TableOfContents {
     createNew: (
       widget: W,
       configuration?: TableOfContents.IConfig
-    ) => IModel<IHeading>;
+    ) => IModel<H>;
   }
 
   /**

